### ISSUE TITLE
UnReferenced irrelevant section causing htmlhelp to fail compiling

### ISSUE THREAD

**Issue by Enigma644** [dita-ot/dita-ot#3047]
## Expected Behavior
HTML Help should compile.
There should be no warnings or errors thrown from content that is not referenced by the root document.

## Actual Behavior
Help CHM does not compile correctly. Internal html files are missing.

## Possible Solution
Topic reader should ignore unreferenced sections.

## Steps to Reproduce
[testFiles.zip](https://github.com/dita-ot/dita-ot/files/2283760/testFiles.zip)

1. Use test content provided in zip
2. Correct paths for your setup
3. Run makeCHM.bat

## Copy of the error message, log file or stack trace
log.log in attached zip.

[topic-reader] file:/C:/Users/admin/Desktop/Output/test/ReferencedTopic.dita:14:59: [DOTJ036W][WARN] The file "file:/C:/Users/admin/Desktop/Output/Irrelevant.dita" is outside the scope of the input dita/map directory.

## Environment

* DITA-OT version: 3.1.1
* Operating system and version: Win10x64
* How did you run DITA-OT? makeCHM.bat included
* Transformation type: htmlhelp

## Notes
ReferencedTopic.dita contains a section "Irrelevant section", which is not referenced by the publishing map. Content in this irrelevant section should not affect anything in this publishing scenario.

The specific problem in this case seems to be that the irrelevant section contains an out of scope link to another dita file (../Irrelevant.dita). If you remove the preceding "../", then the compile still throws an error saying ```Irrelevant.dita' does not exist```, but the CHM is produced correctly.

The use case for this is that the repository does not have file access compatible with the DOTK, so for publishing we are exporting the relevant content to a temporary file system location in order to run the DOTK, however, we do not want to export the entire repository; Only the content necessary for publishing as referenced by the selected map!

**Comment by drmacro** [comment]
The processing shouldn't fail in this case but the markup as provided is very bad practice. All resources used by topics should be reflected in the map, so the ReferencedTopic.dita, which is the conref target of the section in MainTopic.dita should be referenced from the map using a resource-only topicref. Likewise, having an xref to a topic that is not used from the map is a broken link and you should have no expectation that the link will work. In PDF or similar monolithic output the link *cannot* work (because the referenced topic would never be part of the publication.

With DITA 1.2, if you use a key reference, you can have the link redirect to the appropriate target in the context of this map, which might be the topic as published on a Web site or a different topic that is in the map.

With DITA 1.3 you could make the link be a cross-publication link if that's what it is.

**Comment by Enigma644** [comment]
The link is not relevant, it is irrelevant content in the "Irrelevant section" ;-)

None of the content of the "Irrelevant section" in ReferencedTopic.dita is used in this map. So the fact that the link is broken **should be** irrelevant. Only section ReferencedTopic.dita#rts2 is used from MainTopic.dita, which is used by the map. Also, it's not the fact that the link is broken that makes the parser fail, it's only if the link is out of scope. You can leave the broken link in and it will compile the html help just fine, if the broken link is in-scope (i.e. remove the preceding ../ in the path).

map.ditamap -  topicref href="MainTopic.dita"
MainTopic.dita - section conref="ReferencedTopic.dita#rts2"

The "Irrelevant section" ReferencedTopic.dita#rts1 is never used and any errors in it's content should be ignored by the parser as they are irrelevant to the output.

**Comment by drmacro** [comment]
Ah, the link is in a section that is not used from MainTopic.dita.

But there is still the problem that because ReferencedTopic.dita is not explicitly marked as resource-only by a topicref from the map the processor has no way to know for sure what its relevance is or isn't. It is referenced by a conref from MainTopic.dita, which makes the entire topic relevant. The processor may not know at the time it evaluates the links in that topic that it is (as far as the map author is concerned) a resource-only topic because it was never told that. The fact that the section with the link in it is in fact never used is not really something the processor would necessarily know at that point, or it may never bother to know it (that is, the processor is under no obligation to track which things that contain links are or are not actually reflected in the final result so that it can either avoid doing things that would ultimately not need doing (such as checking the resolvability of a link that is never actually rendered) or provide better diagnostic messages.

But note that if you turned on the "only topics in map" flag, the conref would not be resolvable at all.

So again, the failure is bad but your content reflects bad practice that cannot produce reliable results (or desired results) with all transform types. It would be worth the experiment to see if referencing ReferencedTopic.dita with a resource-only topicref changes the failure behavior.

**Comment by robander** [comment]
I'm guessing this comes down to the fact that DITA-OT is very greedy when evaluating links -- it will follow every link it finds, without regard to whether it needs the topic or not. This was a design choice back in DITA-OT -- I don't like it, but it's proven nearly impossible to change without re-architecting pretty much everything. So ... this is an issue that has existed since DITA-OT 1.0. The "only topics in map" flag was a band-aid that tried to make things better, but it is really intended to help control what generates output (it does not control whether links are followed).

With map-first processing introduced in 3.0 it should now be possible to do better, but we're still not there.

Basically if you have a reference to `ReferencedTopic.dita`, and that file has a link to another local topic, DITA-OT is going to try and follow that link. It doesn't matter if you get there by evaluating `@conref` in a topic, or by a normal `<topicref>`, or by a `resource-only` `<topicref>`.

**Comment by robander** [comment]
I can't say for sure -- but HTML Help has long had problems with topics that are outside of the scope of the map directory. Because HTML Help is a zip, where the project files need to be at the root directory, files that end up outside of that directory are missing; moving the zip up to get everything results in a broken CHM. 

I haven't had time to test with these specific files, so I can't say what is causing the compile to fail, but that's my first guess -- that simply trying to follow the link throws off directory detection, and results in compilation running in the wrong spot.

**Comment by Enigma644** [comment]
```outer.control=quiet``` gets rid of the warning, but it still fails with the same ```HHC5003: Error: Compilation failed while compiling MainTopic.html.``` .. so I'm not really sure where the issue is.

**Comment by robander** [comment]
That makes sense -- setting `outer.control=quiet` doesn't actually change any of the processing, all it does is avoid generating the error message. So, nothing about the actual process that led to the failure will change with that. The issue is going to come down to a build variable somewhere getting thrown off between (temp directory) or (directory inside of temp that holds the map) -- those values do not match when there is a `../` link outside of the map dir, and that's where this sort of processing has historically had problems.
